src/graph/Graph.java

```diff
@@ -0,0 +1,172 @@
+package graph;
+
+import java.util.*;
+
+public class Graph {
+
+  private Map<Integer, Vertex> vertices;
+  private String name;
+
+  public Graph() {
+    this("unnamed");
+  }
+
+  public Graph(String name) {
+    this.vertices = new HashMap<>();
+    this.name = name;
+  }
+
+  public void setName(String name) {
+    this.name = name;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public Vertex getNode(int id) {
+    return this.vertices.get(id);
+  }
+
+  public Set<Integer> getNodesID() {
+    return vertices.keySet();
+  }
+
+  public Collection<Vertex> getNodes() {
+    return vertices.values();
+  }
+
+  public Vertex addNode(Vertex v) {
+    this.vertices.put(v.getId(), v);
+    return v;
+  }
+
+  public Vertex addNode(int id) {
+    Vertex node = new Vertex(id);
+    return this.addNode(node);
+  }
+
+  public Vertex addNode() {
+    if (this.isEmpty()) {
+      return this.addNode(0);
+    }
+    int max_id = Collections.max(this.vertices.keySet());
+    return this.addNode(max_id + 1);
+  }
+
+  public void removeNode(int id) {
+    for (Vertex v : this.getNodes()) {
+      v.removeNeighbor(id);
+    }
+    this.vertices.remove(id);
+  }
+
+  public void removeNode(Vertex v) {
+    this.removeNode(v.getId());
+  }
+
+  public Edge addEdge(Vertex node_u, Vertex node_v) {
+    node_v.addNeighbor(node_u);
+    return node_u.addNeighbor(node_v);
```

Comment:
I think you should add the edge object directly into the adjacency list rather than create an Edge object internally, because the Edge's attributes may be useful.